File: tests/test_apps/ruby/comprehensive_test.rb
require 'html_to_markdown'
require 'json'

def load_fixtures(filename)
  fixture_path = File.join(__dir__, '../fixtures', filename)
  JSON.parse(File.read(fixture_path))
end

RSpec.describe 'comprehensive html-to-markdown tests' do
  # Basic HTML conversion tests
  describe 'Basic HTML Conversion' do
    basic_fixtures = load_fixtures('basic-html.json')

    basic_fixtures.each do |test_case|
      it test_case['name'] do
        result = HtmlToMarkdown.convert(test_case['html'], test_case['options'] || {})
        expect(result.strip).to eq(test_case['expectedMarkdown'].strip)
      end
    end
  end

  # Metadata extraction tests
  describe 'Metadata Extraction Feature' do
    it 'converts HTML with metadata extraction enabled' do
      html = '<html><head><title>Test</title></head><body><p>Content</p></body></html>'
      markdown, metadata = HtmlToMarkdown.convert_with_metadata(html)

      expect(markdown).to be_a(String)
      expect(metadata).to be_a(Hash)
      expect(markdown).to include('Content')
    end

    it 'extracts document metadata' do
      html = '<html><head><title>My Page</title><meta name="description" content="A test page"></head><body><p>Content</p></body></html>'
      _, metadata = HtmlToMarkdown.convert_with_metadata(html)

      expect(metadata).to have_key(:document)
      expect(metadata[:document][:title]).to eq('My Page')
      expect(metadata[:document][:description]).to eq('A test page')
    end

    it 'extracts headers from content' do
      html = '<h1>Title</h1><h2>Subtitle</h2><p>Content</p>'
      _, metadata = HtmlToMarkdown.convert_with_metadata(html)

      expect(metadata).to have_key(:headers)
      expect(metadata[:headers]).to be_a(Array)
    end

    it 'extracts links from content' do
      html = '<p><a href="https://example.com">Link</a></p>'
      _, metadata = HtmlToMarkdown.convert_with_metadata(html)

      expect(metadata).to have_key(:links)
      expect(metadata[:links]).to be_a(Array)
    end

    it 'extracts images from content' do
      html = '<p><img src="https://example.com/image.png" alt="Example"></p>'
      _, metadata = HtmlToMarkdown.convert_with_metadata(html)

      expect(metadata).to have_key(:images)
      expect(metadata[:images]).to be_a(Array)
    end
  end

  # Inline images extraction tests
  describe 'Inline Images Feature' do
    it 'returns hash with markdown, inline_images, and warnings' do
      html = '<p><img src="data:image/png;base64,ZmFrZQ==" alt="Test"></p>'
      result = HtmlToMarkdown.convert_with_inline_images(html)

      expect(result).to be_a(Hash)
      expect(result).to have_key(:markdown)
      expect(result).to have_key(:inline_images)
      expect(result).to have_key(:warnings)
    end

    it 'extracts base64-encoded inline images' do
      html = '<p><img src="data:image/png;base64,ZmFrZQ==" alt="Inline image"></p>'
      result = HtmlToMarkdown.convert_with_inline_images(html)

      expect(result[:inline_images]).to be_a(Array)
      unless result[:inline_images].empty?
        expect(result[:inline_images].first).to have_key(:data)
        expect(result[:inline_images].first).to have_key(:description)
      end
    end

    it 'handles multiple inline images' do
      html = '<p><img src="data:image/png;base64,ZmFrZTA=" alt="Image 1"><img src="data:image/png;base64,ZmFrZTE=" alt="Image 2"></p>'
      result = HtmlToMarkdown.convert_with_inline_images(html)

      expect(result[:inline_images]).to be_a(Array)
    end
  end

  # Visitor pattern tests
  describe 'Visitor Pattern Feature' do
    it 'converts HTML with visitor callbacks' do
      html = '<p>Test</p>'
      visitor = double('visitor')
      allow(visitor).to receive(:visit_text).and_return({ type: :continue })
      allow(visitor).to receive(:visit_element_start).and_return({ type: :continue })
      allow(visitor).to receive(:visit_element_end).and_return({ type: :continue })

      result = HtmlToMarkdown.convert_with_visitor(html, nil, visitor)

      expect(result).to be_a(String)
      expect(result).to include('Test')
    end

    it 'supports custom element processing via visitor' do
      html = '<div class="custom">Custom content</div>'
      visitor = double('visitor')
      allow(visitor).to receive(:visit_element_start).and_return({ type: :continue })
      allow(visitor).to receive(:visit_element_end).and_return({ type: :continue })
      allow(visitor).to receive(:visit_text).and_return({ type: :continue })

      result = HtmlToMarkdown.convert_with_visitor(html, nil, visitor)

      expect(result).to be_a(String)
    end
  end

  # Conversion options tests
  describe 'Conversion Options' do
    it 'applies heading style option (atx)' do
      html = '<h1>Title</h1>'
      result = HtmlToMarkdown.convert(html, heading_style: :atx)

      expect(result).to match(/^# Title/)
    end

    it 'applies heading style option (atx_closed)' do
      html = '<h1>Title</h1>'
      result = HtmlToMarkdown.convert(html, heading_style: :atx_closed)

      expect(result).to match(/#.*#/)
    end

    it 'applies list indent width option' do
      html = '<ul><li>Item 1<ul><li>Nested</li></ul></li></ul>'
      result = HtmlToMarkdown.convert(html, list_indent_width: 4)

      expect(result).to be_a(String)
    end

    it 'applies code block style option' do
      html = '<pre><code>code here</code></pre>'
      result = HtmlToMarkdown.convert(html, code_block_style: :fenced)

      expect(result).to include('```')
    end

    it 'applies text wrapping option' do
      long_text = 'A' * 200
      html = "<p>#{long_text}</p>"
      result = HtmlToMarkdown.convert(html, wrap: true, wrap_width: 80)

      expect(result).to be_a(String)
    end
  end

  # Error handling tests
  describe 'Error Handling' do
    it 'handles empty HTML gracefully' do
      result = HtmlToMarkdown.convert('')
      expect(result).to eq('')
    end

    it 'handles malformed HTML gracefully' do
      html = '<p>Unclosed paragraph<div>Nested'
      result = HtmlToMarkdown.convert(html)

      expect(result).to be_a(String)
    end

    it 'handles very large HTML documents' do
      large_html = '<p>' + ('x' * 100_000) + '</p>'
      result = HtmlToMarkdown.convert(large_html)

      expect(result).to be_a(String)
    end

    it 'handles special characters in HTML' do
      html = '<p>&lt;script&gt;alert("xss")&lt;/script&gt;</p>'
      result = HtmlToMarkdown.convert(html)

      expect(result).to be_a(String)
      expect(result).not_to include('<script>')
    end

    it 'handles unicode content' do
      html = '<p>Hello 世界 🌍</p>'
      result = HtmlToMarkdown.convert(html)

      expect(result).to include('世界')
      expect(result).to include('🌍')
    end
  end

  # RBS type definitions validation
  describe 'RBS Type Definitions' do
    it 'HtmlToMarkdown module responds to convert' do
      expect(HtmlToMarkdown).to respond_to(:convert)
    end

    it 'HtmlToMarkdown module responds to convert_with_metadata' do
      expect(HtmlToMarkdown).to respond_to(:convert_with_metadata)
    end

    it 'HtmlToMarkdown module responds to convert_with_inline_images' do
      expect(HtmlToMarkdown).to respond_to(:convert_with_inline_images)
    end

    it 'HtmlToMarkdown module responds to convert_with_visitor' do
      expect(HtmlToMarkdown).to respond_to(:convert_with_visitor)
    end

    it 'HtmlToMarkdown::Options class exists' do
      expect(defined?(HtmlToMarkdown::Options)).to be_truthy
    end
  end
end

File: tests/test_apps/elixir/test/comprehensive_test.exs
defmodule HtmlToMarkdownTestApp.ComprehensiveTest do
  use ExUnit.Case

  @moduledoc """
  Comprehensive test suite for html_to_markdown Hex package.
  Tests basic conversions, error handling, and API behavior.
  """

  defp load_fixtures(filename) do
    path = Path.join([__DIR__, "..", "..", "fixtures", filename])
    case File.read(path) do
      {:ok, content} -> Jason.decode!(content)
      {:error, reason} -> raise "Failed to load fixture #{filename}: #{inspect(reason)}"
    end
  end

  describe "basic HTML conversions" do
    setup do
      {:ok, fixtures: load_fixtures("basic-html.json")}
    end

    test "all basic fixtures convert correctly", %{fixtures: fixtures} do
      Enum.each(fixtures, fn fixture ->
        {:ok, result} = HtmlToMarkdown.convert(fixture["html"], fixture["options"] || %{})
        expected = String.trim(fixture["expectedMarkdown"])
        actual = String.trim(result)

        assert actual == expected, """
        Test case: #{fixture["name"]}
        Expected: #{inspect(expected)}
        Got: #{inspect(actual)}
        """
      end)
    end

    test "basic paragraph conversion" do
      html = "<p>Hello World</p>"
      {:ok, result} = HtmlToMarkdown.convert(html)
      assert String.contains?(result, "Hello World")
    end

    test "basic heading conversion" do
      html = "<h1>Title</h1>"
      {:ok, result} = HtmlToMarkdown.convert(html)
      assert String.starts_with?(result, "# Title")
    end

    test "empty input returns empty string" do
      {:ok, result} = HtmlToMarkdown.convert("")
      assert result == ""
    end

    test "whitespace-only input" do
      {:ok, result} = HtmlToMarkdown.convert("   \n\t  ")
      assert String.trim(result) == ""
    end
  end

  describe "error handling with proper tuple patterns" do
    test "returns {:ok, result} tuple on success" do
      result = HtmlToMarkdown.convert("<p>Test</p>")
      assert is_tuple(result)
      assert tuple_size(result) == 2
      assert {:ok, _content} = result
    end

    test "returns {:error, reason} tuple on failure" do
      # Test with very large invalid HTML that may cause issues
      # Most implementations handle this gracefully, but some may error
      result = HtmlToMarkdown.convert("<")
      case result do
        {:ok, _} -> :ok  # May succeed with lenient parsing
        {:error, _reason} -> :ok  # Expected error case
      end
    end

    test "convert function result is always a tuple" do
      test_cases = [
        "<p>Normal</p>",
        "<h1>Heading</h1>",
        "<div>Nested <span>content</span></div>",
        ""
      ]

      Enum.each(test_cases, fn html ->
        result = HtmlToMarkdown.convert(html)
        assert is_tuple(result), "Expected tuple, got #{inspect(result)}"
        assert tuple_size(result) == 2, "Expected 2-tuple, got #{tuple_size(result)}-tuple"
      end)
    end

    test "successful conversions unwrap with pattern matching" do
      html = "<p>Pattern match test</p>"
      assert {:ok, content} = HtmlToMarkdown.convert(html)
      assert is_binary(content)
      assert String.length(content) > 0
    end
  end

  describe "inline images feature" do
    test "inline images are preserved" do
      html = "<p>Check this <img src=\"/image.png\" alt=\"An image\"/> out</p>"
      {:ok, result} = HtmlToMarkdown.convert(html)
      # Markdown image syntax: ![alt](src)
      assert String.contains?(result, "image") or String.contains?(result, "img")
    end

    test "image with alt text" do
      html = "<img src=\"test.jpg\" alt=\"Test Image\"/>"
      {:ok, result} = HtmlToMarkdown.convert(html)
      # Should contain alt text or image reference
      assert String.length(result) > 0
    end
  end

  describe "metadata extraction" do
    test "extracts basic metadata when available" do
      html = """
      <html>
        <head>
          <title>Page Title</title>
          <meta name="description" content="Page description">
        </head>
        <body><p>Content</p></body>
      </html>
      """
      {:ok, result} = HtmlToMarkdown.convert(html)
      assert is_binary(result)
    end

    test "handles pages without metadata" do
      html = "<p>Just content</p>"
      {:ok, result} = HtmlToMarkdown.convert(html)
      assert is_binary(result)
    end
  end

  describe "complex HTML structures" do
    test "nested lists conversion" do
      html = """
      <ul>
        <li>Item 1</li>
        <li>Item 2
          <ul>
            <li>Nested 1</li>
            <li>Nested 2</li>
          </ul>
        </li>
      </ul>
      """
      {:ok, result} = HtmlToMarkdown.convert(html)
      assert String.contains?(result, "Item 1")
      assert String.contains?(result, "Item 2")
    end

    test "mixed content with various HTML elements" do
      html = """
      <div>
        <h1>Title</h1>
        <p>Paragraph with <strong>bold</strong> and <em>italic</em> text.</p>
        <ul>
          <li>List item 1</li>
          <li>List item 2</li>
        </ul>
        <blockquote>A quote</blockquote>
      </div>
      """
      {:ok, result} = HtmlToMarkdown.convert(html)
      assert String.contains?(result, "Title")
      assert String.contains?(result, "Paragraph")
      assert String.contains?(result, "List item")
    end

    test "code blocks preservation" do
      html = "<pre><code>def hello():\n  print('world')</code></pre>"
      {:ok, result} = HtmlToMarkdown.convert(html)
      assert String.contains?(result, "hello") or String.contains?(result, "print")
    end
  end

  describe "options and configuration" do
    test "accept options as a map" do
      html = "<p>Test</p>"
      options = %{}
      {:ok, result} = HtmlToMarkdown.convert(html, options)
      assert is_binary(result)
    end

    test "handles empty options map" do
      html = "<h1>Title</h1>"
      {:ok, result} = HtmlToMarkdown.convert(html, %{})
      assert String.contains?(result, "Title")
    end
  end

  describe "special characters and encoding" do
    test "handles HTML entities" do
      html = "<p>&lt;div&gt; &amp; special chars</p>"
      {:ok, result} = HtmlToMarkdown.convert(html)
      assert String.length(result) > 0
    end

    test "unicode content preservation" do
      html = "<p>Unicode: 你好 🎉 café</p>"
      {:ok, result} = HtmlToMarkdown.convert(html)
      assert String.length(result) > 0
    end

    test "preserves quotes and apostrophes" do
      html = "<p>He said, \"It's working!\"</p>"
      {:ok, result} = HtmlToMarkdown.convert(html)
      assert String.length(result) > 0
    end
  end

  describe "link handling" do
    test "converts anchor links to markdown" do
      html = "<a href=\"https://example.com\">Example Link</a>"
      {:ok, result} = HtmlToMarkdown.convert(html)
      assert String.contains?(result, "Example Link")
    end

    test "handles links with fragments" do
      html = "<a href=\"#section\">Go to section</a>"
      {:ok, result} = HtmlToMarkdown.convert(html)
      assert String.length(result) > 0
    end
  end

  describe "integration and API contract" do
    test "module is loaded and accessible" do
      assert Code.ensure_loaded?(HtmlToMarkdown)
      assert function_exported?(HtmlToMarkdown, :convert, 1)
      assert function_exported?(HtmlToMarkdown, :convert, 2)
    end

    test "convert/1 and convert/2 both work" do
      html = "<p>Test</p>"
      {:ok, result1} = HtmlToMarkdown.convert(html)
      {:ok, result2} = HtmlToMarkdown.convert(html, %{})
      assert result1 == result2
    end

    test "return values are properly formatted" do
      html = "<strong>Bold</strong>"
      result = HtmlToMarkdown.convert(html)
      assert match?({:ok, _}, result) or match?({:error, _}, result)

      case result do
        {:ok, content} ->
          assert is_binary(content)
          assert not is_nil(content)

        {:error, reason} ->
          assert is_binary(reason) or is_atom(reason)
      end
    end
  end
end
